command/agent/exec/exec_test.go

```diff
@@ -0,0 +1,288 @@
+package exec
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"testing"
+	"time"
+
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/agent/config"
+	"github.com/hashicorp/vault/sdk/helper/logging"
+	"github.com/hashicorp/vault/sdk/helper/pointerutil"
+)
+
+func fakeVaultServer() *httptest.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/v1/kv/my-app/creds", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, `{
+                "request_id": "8af096e9-518c-7351-eff5-5ba20554b21f",
+                "lease_id": "",
+                "renewable": false,
+                "lease_duration": 0,
+                "data": {
+                    "data": {
+                        "password": "s3cr3t",
+                        "user": "app-user"
+                    },
+                    "metadata": {
+                        "created_time": "2019-10-07T22:18:44.233247Z",
+                        "deletion_time": "",
+                        "destroyed": false,
+                        "version": 3
+                    }
+                },
+                "wrap_info": null,
+                "warnings": null,
+                "auth": null
+            }`)
+	})
+
+	return httptest.NewServer(mux)
+}
+
+// TestServer_Run tests various scenarios of using vault agent as a process supervisor
+func TestServer_Run(t *testing.T) {
+	fakeVault := fakeVaultServer()
+	defer fakeVault.Close()
+
+	testCases := map[string]struct {
+		envTemplates                  []*ctconfig.TemplateConfig
+		testAppArgs                   []string
+		testAppStopSignal             os.Signal
+		testAppPort                   int
+		simulatedShutdownWaitDuration time.Duration
+		expected                      map[string]string
+		expectedTestDuration          time.Duration
+		expectedError                 error
+	}{
+		"simple": {
+			envTemplates: []*ctconfig.TemplateConfig{{
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}, {
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.password }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_PASSWORD"),
+			}},
+			testAppArgs:       []string{"--stop-after", "10s"},
+			testAppStopSignal: syscall.SIGTERM,
+			testAppPort:       34001,
```

Comment:
Instead of static ports, could we use something similar to `generateListenerAddress` in `agent_test` to try and avoid collisions?

Reply:
I had a similar concern (port collisions when running the tests from multiple PR's) but if I understand correctly, GHA should run each PR in an isolated/containerized environment so it should not be an issue